helix-term/src/ui/menu.rs

```diff
@@ -58,6 +58,8 @@ pub struct Menu<T: Item> {
     /// (index, score)
     matches: Vec<(usize, i64)>,
 
+    previous_pattern: String,
```

Comment:
Why do we need a previous pattern? Why can't just take from the pattern from the prompt?

Reply:
This is necessary for the `add_options` method, because it needs to recalculate the showed options (`recalculate_score`), without knowing the current pattern.
In the future, this could also help optimizing/filtering previous matches similar as in the picker.

Reply:
Well as said before, the picker already had it in there, so I just adopted this in the menu as well. 
For example `self.matches` is also dependent on the current pattern and stored inside this Component. Actually all state necessary for rendering (or managing state in) the Menu (or Picker) is stored in the Component.

It e.g. makes sense for caching, prefiltering (as done in the picker) etc. 

I think it's maybe better to completely separate data and/or state/model and the view (completion menu in particular), but this is a story(or big PR) for itself, and I'm not sure if it makes the code much better/readable/compact etc.